### ISSUE TITLE
Support for Python 3.12

### ISSUE THREAD

**Issue by davidbrochart** [fastapi-users/fastapi-users#1301]
## Describe the bug

Importing `fastapi_users` fails with Python 3.12.

## To Reproduce

1. Install `fastapi-users` v12.1.2 and run Python 3.12.
2. Execute `import fastapi_users`.
3. See error `ModuleNotFoundError: No module named 'pkg_resources'`.

## Expected behavior

Fastapi-users should work on Python 3.12.

## Configuration
 - Python version : 3.12
 - FastAPI version : 0.103.2
 - FastAPI Users version : 12.1.2

## Additional context

It seems that `passlib` still uses `pkg_resources`, which is deprecated. On Python <3.10, `importlib_metadata` should be used, and `importlib.metadata` should be used on Python >=3.10.

**Comment by frankie567** [comment]
Thanks for raising this @davidbrochart 👍 

Unfortunately, `passlib` seems to be unmaintained those days: https://foss.heptapod.net/python-libs/passlib

It's really unfortunate, because `passlib` is the de-facto standard for password hashing. I'm not aware of alternatives 🤔

**Comment by davidbrochart** [comment]
Maybe fork it?

**Comment by frankie567** [comment]
Probably, but not me. I know nothing about cryptography algorithms, so I wouldn't risk to support a library *that* critical 😅

**Comment by davidbrochart** [comment]
An alternative is to install [setuptools](https://pypi.org/project/setuptools), but there is a deprecation warning when importing:
```
Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pkg_resources
<stdin>:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```

**Comment by tcompa** [comment]
I think that the warning in the previous comment (by @bert2002) is unrelated to python 3.12; it's rather a `passlib` issue with recent `bcrypt` versions (see https://foss.heptapod.net/python-libs/passlib/-/issues/190).

---

> Unfortunately, passlib seems to be unmaintained those days: https://foss.heptapod.net/python-libs/passlib
>
> It's really unfortunate, because passlib is the de-facto standard for password hashing. I'm not aware of alternatives 🤔

For the record, there seems to be a renewed effort towards `passlib` maintenance, see https://foss.heptapod.net/python-libs/passlib/-/issues/187.

**Comment by gaetanDubuc** [comment]
The only dependency to passlib seems to be in **password.py**. https://argon2-cffi.readthedocs.io/en/stable/argon2.html could replace passlib as the package defines a PasswordHasher with close protocol. We just have to define a new `CryptContext` protocol.

**Comment by frankie567** [comment]
> The only dependency to passlib seems to be in **password.py**. https://argon2-cffi.readthedocs.io/en/stable/argon2.html could replace passlib as the package defines a PasswordHasher with close protocol. We just have to define a new `CryptContext` protocol.

Passlib supports several hashing algorithms, not only argon2, and has convenient features like hash upgrades. Plus, I believe most FastAPI Users installations use bcrypt right now.

**Comment by JimScope** [comment]
Hi @frankie567  could this option to unblock the library for python 3.12 be considered as a short term solution? https://github.com/fastapi-users/fastapi-users/pull/1345

**Comment by MatthewScholefield** [comment]
[A passlib maintainer responded](https://foss.heptapod.net/python-libs/passlib/-/issues/187#note_346669) so it sounds like there may be a fix coming soon.

**Comment by frankie567** [comment]
Fixed as of [`v13.0.0`](https://github.com/fastapi-users/fastapi-users/releases/tag/v13.0.0)
